Modules/Module_Processing/Module_ASL/xASL_wrp_ResampleASL.m

```diff
@@ -203,146 +203,180 @@ function xASL_wrp_ResampleASL(x)
-		
-		% For multi-PLD, save both PWI4D and PWI
-		if x.modules.asl.bMultiPLD
-			% Save PWI4D
-			fprintf('%s\n', PathPWI4D{iSpace});
-			xASL_io_SaveNifti(PathASL{iSpace}, PathPWI4D{iSpace}, ASL_im, 32, false);
-		end
-		
-		% For both single- and multi-PLD, create single PWI for further steps in ASL module
-		PWIsingle = xASL_stat_MeanNan(ASL_im,4); % Average across PLDs
+    
+    % =====================================================================
+    % A) Check subtraction
+    if numel(size(ASL_im))>4
+        warning('In BIDS ASL NIfTIs should have [X Y Z n/PLD/TE/etc] as 4 dimensions');
+        error(['NIfTI has more than 4 dimensions: ' PathASL4D{iSpace}]);
         
-		% Save single PWI
-		fprintf('%s\n', PathPWI{iSpace});
-		xASL_io_SaveNifti(PathASL{iSpace}, PathPWI{iSpace}, PWIsingle, 32, false);
-					
-    elseif round(dim4/2)~=dim4/2
-        warning('Odd number of control-label pairs, skipping');
+    elseif dim4>1 && round(dim4/2)~=dim4/2
+        warning('Odd (i.e., not even) number of control-label pairs, skipping');
         return;
         
+        
+    % =====================================================================
+    % B) Time-Encoded subtraction
     elseif x.modules.asl.bTimeEncoded
+        
+        %% 1) Create PWI4D
         % Decoding of TimeEncoded data - it outputs a decoded image and it also saves as a NII
-        ASL_im = xASL_quant_HadamardDecoding(PathASL{iSpace}, x.Q);
+        PWI4D = xASL_quant_HadamardDecoding(PathASL4D{iSpace}, x.Q);
 		
-		% Calculate Hadamard block size (number of unique PLDs * multiTEs) = number of volumes per block
-        vectorSizePLD = length(unique(x.Q.Initial_PLD));
-        blockSize = x.Q.NumberEchoTimes * (vectorSizePLD - (vectorSizePLD/x.Q.TimeEncodedMatrixSize)); 
-		
-        if blockSize ~= size(ASL_im,4) % In case there is more than 1 block, we will average across repetitions
-			if mod(size(ASL_im,4), blockSize)
-				% Number of total volumes cannot be divided by blockSize
-				error(['Total number of volumes ' xASL_num2str(size(ASL_im,4)) ' cannot be composed of blocks of size ' xASL_num2str(blockSize)]);
-			else
-				PWI = zeros(size(ASL_im,1), size(ASL_im,2), size(ASL_im,3), blockSize); % preallocate PWI
-				for iBlock = 1:blockSize
-					PWI(:,:,:,iBlock) = xASL_stat_MeanNan(ASL_im(:,:,:,iBlock:blockSize:end), 4); % Averaged PWI4D across repetitions
-				end
-			end
-        else % if Hadamard Block size == unique PLDs * number of multi-TEs -> a single repetitions -> we don't want to average
-            PWI = ASL_im;
+        %% 2) Create PWI3D
+		% Calculate Hadamard block size (number of unique PLDs.*labeling durations.* echo times) = number of volumes per repetition
+        unique_InitialPLD = unique(x.Q.Initial_PLD);
+        nUnique_InitialPLD = length(unique_InitialPLD);
+
+        nEchoTimes = length(uniquetol(x.EchoTime,0.001)); % Obtain the number of echo times

```

Comment:
True. We need to harmonize the nomenclature, which is what I did here.

Note that I left the complicated echo time definition at the start of `xASL_Module_ASL` for you to check to make it more standardized. Hadamard should not care what is inside its iterations right, you should not need the number of echoes (or other iterations like flip angles, TRs, etc) for calculating things like "block size".

Reply:
Not sure if I really understand your answer. We should indeed keep block-size things internal. But x.Q.NumberEchoTimes should be a useful parameter to know which Echos we have and if we have multi-echo. So that can stay at the start of the ASL module and we should reuse whenever possible. Or not?

Reply:
Yes, we now have x.Q.uniqueEchoTime etc at the start of xASL_module_ASL, so there we can add the x.Q.nUnique* as well